sdk/src/nonce_instruction.rs

```diff
@@ -0,0 +1,479 @@
+use log::*;
+use num_derive::{FromPrimitive, ToPrimitive};
+use serde_derive::{Deserialize, Serialize};
+use solana_logger;
+use crate::{
+    account::{get_signers, KeyedAccount},
+    instruction::{AccountMeta, Instruction, InstructionError},
+    instruction_processor_utils::{limited_deserialize, next_keyed_account, DecodeError},
+    nonce_program::id,
+    nonce_state::{NonceAccount, NonceState},
+    pubkey::Pubkey,
+    system_instruction,
+    sysvar::{
+        Sysvar,
+        recent_blockhashes::{self, RecentBlockhashes},
+        rent::{self, Rent},
+    }
+};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, FromPrimitive, ToPrimitive)]
+pub enum NonceError {
+    NoRecentBlockhashes,
+    NotExpired,
+    UnexpectedValue,
+    BadAccountState,
+}
+
+impl<E> DecodeError<E> for NonceError {
+    fn type_of() -> &'static str {
+        "NonceError"
+    }
+}
+
+impl std::fmt::Display for NonceError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            NonceError::NoRecentBlockhashes => write!(f, "recent blockhash list is empty"),
+            NonceError::NotExpired => write!(f, "stored nonce is still in recent_blockhashes"),
+            NonceError::UnexpectedValue => write!(f, "specified nonce does not match stored nonce"),
+            NonceError::BadAccountState => write!(f, "cannot handle request in current account state"),
+        }
+    }
+}
+
+impl std::error::Error for NonceError {}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub enum NonceInstruction {
+    /// `Nonce` consumes a stored nonce, replacing it with a successor
+    ///
+    /// Expects 2 Accounts:
+    ///     0 - Initialized NonceAccount to be spent
```

Comment:
Looks like `Nonce` is also used to _initialize_ a `NonceAccount` so I think this needs to be updated